lib/child_process.js

```diff
@@ -250,45 +245,43 @@ exports.execFile = function(file /*, args, options, callback*/) {
   }
 
   if (child.stdout) {
-    stdoutState = child.stdout._readableState;
+    if (encoding)
+      child.stdout.setEncoding(encoding);
 
     child.stdout.addListener('data', function(chunk) {
-      // If `child.stdout.setEncoding()` happened in userland, convert string to
-      // Buffer.
-      if (stdoutState.decoder) {
-        chunk = Buffer.from(chunk, stdoutState.decoder.encoding);
-      }
-
-      stdoutLen += chunk.byteLength;
+      stdoutLen += chunk instanceof Buffer ?
+        chunk.length :
+        Buffer.byteLength(chunk, encoding);
 
       if (stdoutLen > options.maxBuffer) {
         ex = new Error('stdout maxBuffer exceeded');
-        stdoutLen -= chunk.byteLength;
         kill();
       } else {
-        _stdout.push(chunk);
+        if (!encoding)
+          _stdout.push(chunk);
+        else
+          _stdout += chunk;
       }
     });
   }
 
   if (child.stderr) {
-    stderrState = child.stderr._readableState;
+    if (encoding)
+      child.stderr.setEncoding(encoding);
 
     child.stderr.addListener('data', function(chunk) {
-      // If `child.stderr.setEncoding()` happened in userland, convert string to
-      // Buffer.
-      if (stderrState.decoder) {
-        chunk = Buffer.from(chunk, stderrState.decoder.encoding);
-      }
-
-      stderrLen += chunk.byteLength;
+      stderrLen += chunk instanceof Buffer ?
```

Comment:
On the other hand, since *all* the code around it seems to check `encoding`, maybe I'll stick with that. 

Changed!